server/src/main/java/org/elasticsearch/common/metrics/Counters.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.metrics;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Simple usage stat counters based on longs. Internally this is a map mapping from String to a CounterMetric. This class also
+ * provides a helper method that converts the counters to a nested map, using the "." as a splitter. This allows the stats producer to not
+ * worry about how the map is actually nested.
+
+ * IMPORTANT: if the consumer of the metrics will make use of the nested map, it is the responsibility of the producer to provide counters
+ * that will not have conflicts, which means that there will be no counter which is a substring of another counter.
```

Comment:
This is a bit confusing, not sure I understand what this means. Maybe because `counter` is overloaded (it's a type that wraps the counters map) and the context here is `substring`-ing the counter?

Reply:
You are right, I will rephrase it